### DISCUSSION TITLE
Add One-way and Multi-way Synonym Support to SeekStorm

### DISCUSSION THREAD

**Discussion by westadaptive** [SeekStorm/SeekStorm#9] [General]
Feature request: I would be really cool if we can have synonym support in SeekStorm for both one way and multi way synonyms if possible.

**Comment by wolfgarbe** [comment]
Yes, this should be easy to implement, if all synonyms are defined at index time.
Dynamically adding/changing synonyms **after** documents were indexed is more complex and expensive concerning latency.

Supporting phrase synonyms like "smartphone" = "smart phone" = "mobile phone"  or "personal computer" = "pc" would be probably also desirable.

**Reply by westadaptive** [reply]
This would be great at index time, I would be happy to test on an address database as all of the synonmys will be static for prefix and suffixes for streets. Example : [Avenue, Ave, Av ] , [Boulevard, Blvd, Bl] .

**Comment by wolfgarbe** [comment]
One-way and multi-way synonym support implemented. 
New synonyms parameter added to **create index** REST API endpoint (and create_index library method):

`curl.exe --request POST --url 'http://127.0.0.1/api/v1/index' --header 'apikey: 
AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=' --header 'content-type: application/json' --data '{ \"schema\": [{\"field\": \"postcode\", \"field_type\": \"Text\",\"stored\": true, \"indexed\": true,\"boost\":10.0},{\"field\": \"address\", \"field_type\": \"Text\",\"stored\": true, \"indexed\": true}], \"index_name\": \"test_index\",\"similarity\": \"Bm25fProximity\",\"tokenizer\": \"UnicodeAlphanumeric\",\"synonyms\":[{\"terms\": [\"avenue\", \"ave\", \"av\"]},{\"terms\": [\"boulevard\", \"blvd\", \"bl\"]}]}'`


<img width="524" alt="image" src="https://github.com/user-attachments/assets/e1d52c7f-6376-4e8a-b551-ea354921894a">

Need to do some more tests, documentation and add synonym support to highlighting before release of v0.10.0.

**Reply by westadaptive** [reply]
Thank you so much, this is awesome! Looking forward to it.

**Comment by wolfgarbe** [comment]
[SeekStorm v0.10.0](https://github.com/SeekStorm/SeekStorm/releases/tag/v0.10.0) has been released: one-way and multi-way synonym definition per index added

![image](https://github.com/user-attachments/assets/d6f3f708-86d8-40fd-90d7-99026110788d)

**Comment by westadaptive** [comment]
This is amazing, Thank you!!!
